ir/js/src/main/scala/org/scalajs/ir/SHA1.scala

```diff
@@ -0,0 +1,283 @@
+       */
+      var i = 0
+      while (i != 16) {
+        W(i) = bufferView.getInt32(i * 4)
+        i += 1
+      }
+      while (i != 80) {
+        W(i) = rol(W(i - 3) ^ W(i - 8) ^ W(i - 14) ^ W(i - 16), 1)
+        i += 1
+      }
+
+      // Copy state to working vars
+      var a = state0
+      var b = state1
+      var c = state2
+      var d = state3
+      var e = state4
+
+      // Main loop
+
+      @inline def performOneIteration(fi: Int, Ki: Int): Unit = {
+        val temp = rol(a, 5) + fi + e + Ki + W(i)
+        e = d
+        d = c
+        c = rol(b, 30)
+        b = a
+        a = temp
+      }
+
+      i = 0
+      while (i != 20) {
+        performOneIteration((b & c) | (~b & d), 0x5a827999)
+        i += 1
+      }
+      while (i != 40) {
+        performOneIteration(b ^ c ^ d, 0x6ed9eba1)
+        i += 1
+      }
+      while (i != 60) {
+        performOneIteration((b & c) | (b & d) | (c & d), 0x8f1bbcdc)
+        i += 1
+      }
+      while (i != 80) {
+        performOneIteration(b ^ c ^ d, 0xca62c1d6)
+        i += 1
+      }
+
+      // Add the working vars back into `state`
+      state0 += a
+      state1 += b
+      state2 += c
+      state3 += d
+      state4 += e
+    }
+
+    private def sha1Update(data: Array[Byte], off: Int, len: Int): Unit = {
+      // Local copies
+      val buffer = this.buffer
+
+      /* Update the byte count.
+       * We use `Integer.toUnsignedLong(len)` because `len` is known to be
+       * non-negative, and it's faster. It also results in a `hi` part which is
+       * a constant 0, which makes the `+` operation faster as well.
+       */
+      val oldByteCount = byteCount
+      byteCount = oldByteCount + Integer.toUnsignedLong(len)
+
+      /* Decompose `data` into 64-byte chunks, taking into account bytes that
+       * were previously stored in `buffer`.
+       */
+
+      val alreadyInBufferCount = oldByteCount.toInt & 63
+
+      if (alreadyInBufferCount + len < 64) {
+        // We won't fill the buffer even once. Just copy `data` to the buffer.
+        arraycopyToInt8Array(data, off, buffer, alreadyInBufferCount, len)
+      } else {
+        // We will fill the buffer at least once
+
+        var currentDataPos = off
+        val endInData = off + len
+
+        // Fill the buffer once, and transform
+        val toCompleteFirstChunk = 64 - alreadyInBufferCount
+        arraycopyToInt8Array(data, currentDataPos, buffer, alreadyInBufferCount,
+            toCompleteFirstChunk)
+        sha1Transform()
+        currentDataPos += toCompleteFirstChunk
+
+        // Transform more complete chunks
+        while (currentDataPos + 64 <= endInData) {
+          arraycopyToInt8Array(data, currentDataPos, buffer, 0, 64)
+          sha1Transform()
+          currentDataPos += 64
+        }
+
+        // Copy the remaining data in the buffer, to be used next time
+        arraycopyToInt8Array(data, currentDataPos, buffer, 0,
+            endInData - currentDataPos)
+      }
```

Comment:
By the way, it seems to me this would be simpler if we used a `ByteBuffer`:

```scala
var dataPos = off
val dataEnd = off + len

while (dataEnd - dataPos >= buffer.remaining()) {
  val c = buffer.remaining()
  buffer.put(data, dataPos, c)
  dataPos += c
  sha1Transform() // flips and clears buffer.
}

buffer.put(data, dataPos, dataEnd - dataPos)
```

Reply:
A `ByteBuffer` might make this portion of the code slightly clearer. But it comes at a significant performance cost because of the abstraction it provides (checks for read-only + bounds-checks + endianness testing). For JS-only code like this, IMO directly using typed arrays is nicer.